Add a functionality to create a snapshot before a disk is deleted

karolgorc opened the pull request:
closes https://github.com/hashicorp/terraform-provider-google/issues/15103
<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->

**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:enhancement
compute: added `create_snapshot_before_destroy` to `google_compute_disk` and `google_compute_region_disk` to enable creating a snapshot before disk deletion
```

```release-note:enhancement
compute: added `rsa_encrypted_key` to `google_compute_region_disk`
```

melinath commented:
@rileykarson are you able to review this?

rileykarson commented:
Been low on cycles, rolled @c2thorn as an alternate reviewer

karolgorc commented:
Yes, sure. One question tough. The snapshot that's created is outside of terraform so it won't be deleted after the test finishes. I'm guessing that we don't want to leave artifacts so what i can do is add a delete API query as a check step. Is that good enough?

Also just remembered that i need to add `sourceDiskEncryptionKey` support into the code here so it will work with encrypted disks.

c2thorn commented:
> Yes, sure. One question tough. The snapshot that's created is outside of terraform so it won't be deleted after the test finishes. I'm guessing that we don't want to leave artifacts so what i can do is add a delete API query as a check step. Is that good enough?

Good point, yes that would be great in a check step.

rileykarson commented on mmv1/products/compute/Disk.yaml:
```diff
@@ -514,3 +514,9 @@ properties:
+    type: Boolean
+    default_value: false
+    description: |
+      If set to true, a snapshot of the disk will be created before it is destroyed.
```
drive-by comment: Can we document the naming format here, and can we consider how we handle long disk names that will go past the name size limit? We may want a `create_snapshot_before_destroy_name_prefix` field for complex cases.

karolgorc replied:
Added

karolgorc commented:
- Improved documentation
- Added missing fields to region disk so the functionality between disk and region_disk will be the same
- Added tests with all encryption methods and disk naming tested
- Added `create_snapshot_before_destroy_prefix`

modular-magician commented:
#### Tests analytics
Total tests: 471
Passed tests: 409
Skipped tests: 61
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccComputeDisk_createSnapshotBeforeDestroy
#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeRegionDisk_createSnapshotBeforeDestroy</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccComputeRegionDisk_createSnapshotBeforeDestroy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/964f19a8-f0dc-42f0-b98b-4099557b9f84/recording/TestAccComputeRegionDisk_createSnapshotBeforeDestroy.log)]


🔴 Tests failed when rerunning REPLAYING mode:
`TestAccComputeRegionDisk_createSnapshotBeforeDestroy` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/964f19a8-f0dc-42f0-b98b-4099557b9f84/build-log/replaying_build_after_recording/TestAccComputeRegionDisk_createSnapshotBeforeDestroy_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/964f19a8-f0dc-42f0-b98b-4099557b9f84/replaying_after_recording/TestAccComputeRegionDisk_createSnapshotBeforeDestroy.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/964f19a8-f0dc-42f0-b98b-4099557b9f84/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/964f19a8-f0dc-42f0-b98b-4099557b9f84/recording) for each test

c2thorn commented on mmv1/third_party/terraform/services/compute/resource_compute_disk_test.go.tmpl:
```diff
@@ -956,6 +977,51 @@ func TestAccComputeDisk_featuresUpdated(t *testing.T) {
 }
 
+func TestAccComputeDisk_createSnapshotBeforeDestroy(t *testing.T) {
+	t.Parallel()
```
Can we skip VCR with 
```suggestion	
	acctest.SkipIfVcr(t) // Disk cleanup test check 
	t.Parallel()
```

c2thorn reviewed (changes requested):
after we remove from VCR I will reconfirm with a manual test run in teamcity

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 404 insertions(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 404 insertions(+), 1 deletion(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12947-old..auto-pr-12947) ( 1 file changed, 11 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 1107
Passed tests: 1030
Skipped tests: 76
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccComputeDisk_createSnapshotBeforeDestroy
#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeRegionDisk_createSnapshotBeforeDestroy</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccComputeRegionDisk_createSnapshotBeforeDestroy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/e5f99552-c0c6-4da3-9ba9-a334fea472c0/recording/TestAccComputeRegionDisk_createSnapshotBeforeDestroy.log)]


🔴 Tests failed when rerunning REPLAYING mode:
`TestAccComputeRegionDisk_createSnapshotBeforeDestroy` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/e5f99552-c0c6-4da3-9ba9-a334fea472c0/build-log/replaying_build_after_recording/TestAccComputeRegionDisk_createSnapshotBeforeDestroy_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/e5f99552-c0c6-4da3-9ba9-a334fea472c0/replaying_after_recording/TestAccComputeRegionDisk_createSnapshotBeforeDestroy.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/e5f99552-c0c6-4da3-9ba9-a334fea472c0/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/e5f99552-c0c6-4da3-9ba9-a334fea472c0/recording) for each test

karolgorc commented:
Tested this just to be sure

```
--- PASS: TestAccComputeRegionDisk_createSnapshotBeforeDestroy (86.15s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google/services/compute  86.180s
```

c2thorn commented on mmv1/third_party/terraform/services/compute/resource_compute_region_disk_test.go.tmpl:
```diff
@@ -252,6 +252,50 @@ func TestAccComputeRegionDisk_featuresUpdated(t *testing.T) {
 }
 
+func TestAccComputeRegionDisk_createSnapshotBeforeDestroy(t *testing.T) {
+	t.Parallel()
```
```suggestion
	acctest.SkipIfVcr(t) // Disk cleanup test check
	t.Parallel()
```

c2thorn commented:
Forgot to skip the region version too

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 405 insertions(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 405 insertions(+), 1 deletion(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12947-old..auto-pr-12947) ( 1 file changed, 11 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 1107
Passed tests: 1030
Skipped tests: 77
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccComputeDisk_createSnapshotBeforeDestroy
- TestAccComputeRegionDisk_createSnapshotBeforeDestroy
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/48bbf931-9a30-4163-a561-4f37d54dec08/build-log/replaying_test.log)

c2thorn commented:
running manual compute disk and region disk tests on TC

karolgorc commented:
Is there anything that has to be done with this?

c2thorn commented:
> Is there anything that has to be done with this?

Apologies, I was out of office for a bit, but just confirmed both the regional and non regional tests have passed. Unfortunately within that timeframe it seems there is now a conflict in the template. Once we resolve that this should be good to go.

c2thorn commented:
I've resolved the conflict in the web editor and will confirm before approving.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 450 insertions(+), 32 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 438 insertions(+), 32 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12947-old..auto-pr-12947) ( 1 file changed, 11 insertions(+))



## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_compute_disk` (139 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_disk" "primary" {
  create_snapshot_before_destroy        = # value needed
  create_snapshot_before_destroy_prefix = # value needed
}


```

modular-magician commented:
#### Tests analytics
Total tests: 0
Passed tests: 0
Skipped tests: 0
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccComputeDisk_createSnapshotBeforeDestroy
- TestAccComputeRegionDisk_createSnapshotBeforeDestroy
🔴 Errors occurred during REPLAYING mode. Please fix them to complete your PR.

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/daca6700-fb31-4ccc-b04a-47556f05276d/build-log/replaying_test.log)

karolgorc commented:
One of the tests didn't have it's brackets closed after resolving the conflicts

```
E0227 07:13:50.806201     485 template_data.go:231] error formatting /home/test/go/src/github.com/hashicorp/terraform-provider-google/google/services/compute/resource_compute_disk_test.go: 1949:26: missing ',' before newline in argument list (and 9 more errors)
```

Fixed

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 403 insertions(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12947-old..auto-pr-12947) ( 8 files changed, 403 insertions(+), 1 deletion(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12947-old..auto-pr-12947) ( 1 file changed, 11 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 1122
Passed tests: 1043
Skipped tests: 79
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccComputeDisk_createSnapshotBeforeDestroy
- TestAccComputeRegionDisk_createSnapshotBeforeDestroy
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12947/artifacts/ea7d59ab-c275-44a8-bdf9-7c895e1a7baa/build-log/replaying_test.log)

c2thorn commented:
> One of the tests didn't have it's brackets closed after resolving the conflicts
> 
> ```
> E0227 07:13:50.806201     485 template_data.go:231] error formatting /home/test/go/src/github.com/hashicorp/terraform-provider-google/google/services/compute/resource_compute_disk_test.go: 1949:26: missing ',' before newline in argument list (and 9 more errors)
> ```
> 
> Fixed

thanks for the fix!

c2thorn reviewed (approved):
updated the changelog entries. LGTM
